### PR TITLE
Shape Detection: Add Callback Mechanism

### PR DESCRIPTION
**Pull request by sgiraudot**
## Summary of Changes

[Pre-approved small feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Shape_Detection_Callbacks) that introduces a callback mechanism to Shape Detection.

## Release Management

* Affected package(s): Point Set Shape Detection, STL Extension (`cpp11::function`)
* Feature/Small Feature (if any): [Features/Small_Features/Shape_Detection_Callbacks](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Shape_Detection_Callbacks)
* pre-approved --Andreas Fabri 15:22, 20 December 2017 (CET)

### CODE REVIEW

File: Point_set_shape_detection_3/include/CGAL/Shape_detection_3/Region_growing.h [hunk 1]
```diff
@@ -169,7 +170,7 @@ shape. The implementation follows \cgalCite{cgal:lm-clscm-12}.
 
     
   private:
-
+
```
Comment (sloriot) [inline Point_set_shape_detection_3/include/CGAL/Shape_detection_3/Region_growing.h]: extra space

File: STL_Extension/doc/STL_Extension/CGAL/function.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
+namespace CGAL {
```
Comment (sloriot) [inline STL_Extension/doc/STL_Extension/CGAL/function.h]: license header!

**Comment by sgiraudot** [comment]
This PR was rebase on https://github.com/CGAL/cgal/pull/2694 and the related deprecated warning was fixed.

**Comment by sgiraudot** [comment]
(Travis failed because of `apt-get`.)

**Comment by lrineau** [comment]
1. Somebody has restarted the failed Travis job.
2. There is a tiny conflict with `master` because of commit 8fc5756cc3f5b10db61ed06805bdde023a41b6a4.

**Comment by sgiraudot** [comment]
Rebased.

**Comment by lrineau** [comment]
Tested in [CGAL-4.12-Ic-142].

  [CGAL-4.12-Ic-142]: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-142.shtml

**Pull request merged**